common/persistence/visibility_hybrid_manager.go

```diff
@@ -0,0 +1,554 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	visibilityHybridManager struct {
+		logger                    log.Logger
+		visibilityMgrs            map[string]VisibilityManager
+		readVisibilityStoreName   dynamicconfig.StringPropertyFnWithDomainFilter
+		writeVisibilityStoreName  dynamicconfig.StringPropertyFn
+		logCustomerQueryParameter dynamicconfig.BoolPropertyFnWithDomainFilter
+	}
+)
+
+const (
+	ContextKey           = ResponseComparatorContextKey("visibility-override")
+	dbVisStoreName       = "db"
+	advancedWriteModeOff = "off"
+)
+
+// ResponseComparatorContextKey is for Pinot/ES response comparator. This struct will be passed into ctx as a key.
+type ResponseComparatorContextKey string
+
+type OperationType string
+
+var Operation = struct {
+	LIST  OperationType
+	COUNT OperationType
+}{
+	LIST:  "list",
+	COUNT: "count",
+}
+
+var _ VisibilityManager = (*visibilityHybridManager)(nil)
+
+// NewVisibilityTripleManager create a visibility manager that operate on DB or advanced visibility based on dynamic config.
+// For Pinot migration, Pinot is the destination visibility manager, ES is the source visibility manager, and DB is the fallback.
+// For OpenSearch migration, OS is the destination visibility manager, ES is the source visibility manager, and DB is the fallback.
+func NewVisibilityHybridManager(
+	visibilityMgrs map[string]VisibilityManager,
+	readVisibilityStoreName dynamicconfig.StringPropertyFnWithDomainFilter,
+	writeVisibilityStoreName dynamicconfig.StringPropertyFn,
+	logCustomerQueryParameter dynamicconfig.BoolPropertyFnWithDomainFilter,
+	logger log.Logger,
+) VisibilityManager {
+	if len(visibilityMgrs) == 0 {
+		logger.Fatal("require one of dbVisibilityManager or advancedVisibilityManager")
```

Comment:
change the error message to match the validation. there's no dbVisibilityManager or advancedVisibilityManager

Reply:
Updated!